src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -74,13 +106,15 @@ def resolve(self, root_reqs, check_supported_wheels):
         requirements = []
         for req in root_reqs:
             if req.constraint:
-                assert req.name
-                assert req.specifier
+                # Ensure we only accept valid constraints
+                reject_invalid_constraint_types(req)
+
+                specifier = req.specifier or SpecifierSet()
```

Comment:
I think `req.specifier` is guarenteed to be a `SpecifierSet` (even for a URL requirement, in which case it would be an empty `SpecifierSet()`, so the `or` part is redundant. If we want to make sure the specifier exist, I’d add an assertion here instead and call it a `packaging` bug if that ever fails.

_Originally posted by @uranusjr in https://github.com/pypa/pip/pull/8239/files_

Reply:
Whoops, indeed - I missed this on my pass through. Thanks for the reminder. I'll fix it.

Reply:
I didn't bother with an assert, I don't think it's needed.